edb/server/protocol/binary.pyx

```diff
@@ -2048,10 +2049,19 @@ cdef class EdgeConnection:
 
         # number of elements in the tuple
         # for empty tuple it's okay to send zero-length arguments
-        # TODO(tailhook) force zero-length arguments for protocol 0.12 after
-        # upgrading the bindings
-        if frb_get_len(&in_buf) == 0:
-            recv_args = 0
+        if self.protocol_version >= (0, 12):
+            frb_len = frb_get_len(&in_buf)
+            is_null_type = \
+                query.query_unit.in_type_id == sertypes.NULL_TYPE_ID.bytes
+            if frb_len and is_null_type:
+                raise errors.ProtocolError(
+                    "absence of query arguments must be encoded with a 'zero' "
+                    "type (id: 00000000-0000-0000-0000-000000000000, "
+                    "encoded with zero bytes)")
+            elif frb_len == 0:
+                recv_args = 0
```

Comment:
Hm, you should check if it's `not is_null_type` here and throw an error. You should have 4 states total for `frb_len == 0`, `frb_len > 0`, `is_null_type == true` and `is_null_type == false` and only `is_null_type && not frb_len` is valid out of those.

Reply:
changed

Reply:
Please add `assert is_null_type` here for clarity.

Reply:
changed